firmware/common2015/utils/logger/logger.hpp

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <cstdarg>
+#include <string>
+
+#include "rj-macros.hpp"
+
+// Do weird macro things for logging the filename and line for every call.
+// Also allows for disabling all logging through macros so all log calls can be removed from production builds.
+#ifdef RJ_LOGGING_EN
+
+#ifdef __FILE_NAME__
+#define __BASE_FILE_NAME__  __FILE_NAME__
+#else   // __FILE_NAME__
+#define __BASE_FILE_NAME__  __FILE__
+#endif
+
+#define GEN_LOG_STR(x,y)    x ":" TO_STRING(y)
+#define LOG_LINE            GEN_LOG_STR(__BASE_FILE_NAME__, __LINE__)
+#define LOG(lvl, ...)       log(lvl, LOG_LINE, __func__, __VA_ARGS__)
+
+#else   // RJ_LOGGING_EN
+#define LOG(...)            // Nothing
+#endif
+
+
+#define FOREACH_LEVEL(LEVEL)    LEVEL(LOG_LEVEL_START)      \
```

Comment:
Ok, makes sense.  One thing I've seen done before (especially in python) that's kinda neat is passing a separate `Logger` instance to different modules.  So in main(), we could pass a `Logger` to the radio module, one to the fpga module, etc.  In debug mode, we'd have all of them set the log level to INFO, but in production, we'd probably bump it up to WARN.  The nice thing with this is that logging level isn't global.  That way once radio is a solved problem, you can set it's logger to WARN, but turn the fpga logger to INFO while you work on it.  Not something for this pull request, but maybe something to consider for later.

Also, @barulicm added some things to the logging utils on the master branch including stream-based logging (like cout rather than printf) that I think got lost in the merge.  Take a look at the diff to see the master version and see if you can retrieve the new bits from that.